optaplanner-examples/src/main/java/org/optaplanner/examples/common/experimental/impl/ConsecutiveIntervalInfoImpl.java

```diff
@@ -0,0 +1,214 @@
+
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.NavigableSet;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.function.BiFunction;
+
+import org.apache.commons.lang3.ObjectUtils;
+import org.optaplanner.examples.common.experimental.api.ConsecutiveIntervalInfo;
+import org.optaplanner.examples.common.experimental.api.IntervalBreak;
+import org.optaplanner.examples.common.experimental.api.IntervalCluster;
+
+public class ConsecutiveIntervalInfoImpl<IntervalType_, PointType_ extends Comparable<PointType_>, DifferenceType_ extends Comparable<DifferenceType_>>
+        implements
+        ConsecutiveIntervalInfo<IntervalType_, PointType_, DifferenceType_> {
+    private final NavigableMap<IntervalSplitPoint<IntervalType_, PointType_>, IntervalClusterImpl<IntervalType_, PointType_, DifferenceType_>> clusterStartSplitPointToCluster;
+    private final NavigableSet<IntervalSplitPoint<IntervalType_, PointType_>> splitPointSet;
+    private final NavigableMap<IntervalSplitPoint<IntervalType_, PointType_>, IntervalBreakImpl<IntervalType_, PointType_, DifferenceType_>> clusterStartSplitPointToNextBreak;
+    private final Iterable<IntervalCluster<IntervalType_, PointType_, DifferenceType_>> intervalClusterIterable;
+    private final BiFunction<PointType_, PointType_, DifferenceType_> differenceFunction;
+    private final Iterable<IntervalBreak<IntervalType_, PointType_, DifferenceType_>> breaksIterable;
+
+    public ConsecutiveIntervalInfoImpl(TreeSet<IntervalSplitPoint<IntervalType_, PointType_>> splitPointSet,
+            BiFunction<PointType_, PointType_, DifferenceType_> differenceFunction) {
+        clusterStartSplitPointToCluster = new TreeMap<>();
+        clusterStartSplitPointToNextBreak = new TreeMap<>();
+        intervalClusterIterable = new MapValuesIterable<>(clusterStartSplitPointToCluster);
+        breaksIterable = new MapValuesIterable<>(clusterStartSplitPointToNextBreak);
+        this.splitPointSet = splitPointSet;
+        this.differenceFunction = differenceFunction;
+    }
+
+    protected void addInterval(Interval<IntervalType_, PointType_> interval) {
+        NavigableMap<IntervalSplitPoint<IntervalType_, PointType_>, IntervalClusterImpl<IntervalType_, PointType_, DifferenceType_>> intersectedIntervalClusterMap =
+                clusterStartSplitPointToCluster.subMap(
+                        ObjectUtils.defaultIfNull(clusterStartSplitPointToCluster.floorKey(interval.getStartSplitPoint()),
+                                interval.getStartSplitPoint()),
+                        true, interval.getEndSplitPoint(), true);
+
+        // Case: the interval cluster before this interval does not intersect this interval
+        if (!intersectedIntervalClusterMap.isEmpty()
+                && intersectedIntervalClusterMap.get(intersectedIntervalClusterMap.firstKey()).getEndSplitPoint()
+                        .compareTo(interval.getStartSplitPoint()) < 0) {
```

Comment:
With these long names and long chains of getters, it becomes hard to track what's going on here. In such cases, it doesn't hurt to help humans a bit. 

There are many ways how to do that. For example, this could be its own boolean variable, with a well chosen name:

     intersectedIntervalClusterMap.get(intersectedIntervalClusterMap.firstKey()).getEndSplitPoint()
                        .compareTo(interval.getStartSplitPoint()) < 0

That will then make the condition a nicely readable one-liner. Staying with the same code, I would consider:

- extracting `intersectedIntervalClusterMap.get(intersectedIntervalClusterMap.firstKey())` to its own well-named variable;
- introducing `isAfter(..)`/`isBefore(...)` on `SplitPoint`, so that all the nasty comparisons are hidden away.

The goal of all of that is to reduce the complexity of the code, so that it reads better. Remember: code is written once, and then read a hundred times over its lifetime.

Reply:
Can't do extract `intersectedIntervalClusterMap.get(intersectedIntervalClusterMap.firstKey())` into its own variable since `intersectedIntervalClusterMap` in the first if statement may be empty. It is extracted into intervalCluster (now `firstIntersectedIntervalCluster`) variable later on.

Reply:
(isAfter/isBefore methods are added on SplitPoint)